src/seccomp_fd_notify.c

```diff
@@ -0,0 +1,164 @@
+
+/* For the x32 ABI, all system call numbers have bit 30 set */
+
+#define X32_SYSCALL_BIT         0x40000000
+
+#define X86_64_CHECK_ARCH_AND_LOAD_SYSCALL_NR \
+	BPF_STMT(BPF_LD | BPF_W | BPF_ABS, \
+		(offsetof(struct seccomp_data, arch))), \
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, AUDIT_ARCH_X86_64, 0, 2), \
+	BPF_STMT(BPF_LD | BPF_W | BPF_ABS, \
+		 (offsetof(struct seccomp_data, nr))), \
+	BPF_JUMP(BPF_JMP | BPF_JGE | BPF_K, X32_SYSCALL_BIT, 0, 1), \
+	BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_KILL_PROCESS)
+
+/* installNotifyFilter() installs a seccomp filter that generates user-space
+   notifications (SECCOMP_RET_USER_NOTIF) when the process calls mkdir(2); the
+   filter allows all other system calls.
+
+   The function return value is a file descriptor from which the user-space
+   notifications can be fetched. */
+
+static int installNotifyFilter(void) {
+	struct sock_filter filter[] = {
+		X86_64_CHECK_ARCH_AND_LOAD_SYSCALL_NR,
+
+		/* Trap some syscalls */
+		BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, __NR_bpf, 0, 1),
+		BPF_STMT(BPF_RET + BPF_K, SECCOMP_RET_USER_NOTIF),
+		BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, __NR_perf_event_open, 0, 1),
+		BPF_STMT(BPF_RET + BPF_K, SECCOMP_RET_USER_NOTIF),
+		BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, __NR_connect, 0, 1),
+		BPF_STMT(BPF_RET + BPF_K, SECCOMP_RET_USER_NOTIF),
+
+		/* Every other system call is allowed */
+		BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
+	};
+
+	struct sock_fprog prog = {
+		.len = (unsigned short) (sizeof(filter) / sizeof(filter[0])),
+		.filter = filter,
+	};
+
+	/* Install the filter with the SECCOMP_FILTER_FLAG_NEW_LISTENER flag; as
+	   a result, seccomp() returns a notification file descriptor. */
+
+	/* Only one listening file descriptor can be established. An attempt to
+	   establish a second listener yields an EBUSY error. */
+
+	/* Because of conflicting uses of the system call return value,
+	   'flags' can't contain both SECCOMP_FILTER_FLAG_NEW_LISTENER
+	   and SECCOMP_FILTER_FLAG_TSYNC. */
+
+
+	int notifyFd = seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_NEW_LISTENER, &prog);
+	if (notifyFd == -1) {
+		PRINT_WARNING("seccomp-install-notify-filter failed");
+		return -1;
+	}
+
+	PRINT_INFO("seccomp notify fd returned %d", notifyFd);
+	return notifyFd;
+}
+
+void maybe_setup_seccomp_notifer() {
+	char *socket_path;
+	socket_path = getenv(TITUS_SECCOMP_NOTIFY_SOCK_PATH);
+	if (socket_path) {
+
+		int sockfd = -1;
+		sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
+		if (sockfd == -1) {
+			PRINT_WARNING("Unable to open unix socket for seccomp handoff: '%s'", strerror(errno));
+			return;
+		}
+	
+		struct sockaddr_un addr = { 0 };
+		addr.sun_family = AF_UNIX;
+		strncpy(addr.sun_path, socket_path, sizeof(addr.sun_path)-1);
+		if (connect(sockfd, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
+			PRINT_WARNING("Unable to connect on unix socket (%s) for seccomp handoff: '%s'", socket_path, strerror(errno));
+			return;
```

Comment:
shouldn't this return an error and be an int

Reply:
I didn't want to this to be be failing condition this early on. There are so many things that could go wrong, and I don't want a lack of BPF to be in the critical path of a task failing.

Reply:
Maybe not this one, but if sendmsg fails for example, that could indicative something else is broke?

Reply:
send_fd does return an error code. `sendmsg`'s return code is now also inspected.

I just wanted to make sure this `maybe_setup_seccomp_notifer` function was always bulletproof and could never crash a task.